sql: remove filter from scanNode

RaduBerinde opened the pull request:
Remove the filter from scanNode. It doesn't provide any benefit
anymore and it just causes confusion (e.g. soft limit handling). In
addition, we want EXPLAIN to map to exec.Factory calls as much as
possible; in this case we want the filter to be shown as a separate
node.

Release note (sql change): EXPLAIN (PLAN) now shows any filter on a
scan as a separate `filter` node.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51493)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 34 of 34 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51493#-:-MCNPAA3BM3mhTbJze-n:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/distsql_physical_planner.go, line 471 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51493#-MCNOYHS94xhYpsY9UOQ:-MCNOYHS94xhYpsY9UOR:b-5ok2di) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cd44e9a1da9508e7b8e23f93d9f012ca7366206/pkg/sql/distsql_physical_planner.go#L471)):*
> ```Go
> 		rec := canDistribute
> 		// Check if we are doing a full scan.
> 		if n.isFull {
> ```

I wonder whether we should change the recommendation of the scan distribution when performing partial scan, probably based on the estimated row count and some threshold setting, what do you think? I guess it is orthogonal to this PR, so just thinking out loud.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51493)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 471 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51493#-MCNOYHS94xhYpsY9UOQ:-MCN_Jgs-9udujZd0Ukm:bc0ttjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cd44e9a1da9508e7b8e23f93d9f012ca7366206/pkg/sql/distsql_physical_planner.go#L471)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I wonder whether we should change the recommendation of the scan distribution when performing partial scan, probably based on the estimated row count and some threshold setting, what do you think? I guess it is orthogonal to this PR, so just thinking out loud.
</blockquote></details>

Yeah I'd expect it to be a threshold on the estimated row count


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: remove filter from scanNode

Remove the filter from scanNode. It doesn't provide any benefit
anymore and it just causes confusion (e.g. soft limit handling). In
addition, we want EXPLAIN to map to exec.Factory calls as much as
possible; in this case we want the filter to be shown as a separate
node.

Release note (sql change): EXPLAIN (PLAN) now shows any filter on a
scan as a separate `filter` node.

RaduBerinde commented:
bors r+
